### PR TITLE
perf: upgraded to yarn 4

### PR DESCRIPTION
**Pull request by alper98**
fixes: https://github.com/backstage/backstage/issues/23327

This pull request includes changes to upgrade the Yarn version from 3 to 4 in the project. The most important changes include updating the Yarn version in the `.yarnrc.yml` and `package.json` files, and modifying the Yarn migration guide in the `yarn-migration.md` file.

Yarn version upgrade:

- As mentioned in the release notes of [yarn4](https://yarnpkg.com/blog/release/4.0) all official plugins are now included per default, therefore `@yarnpkg/plugin-workspace-tools` have been removed

- The binary (cjs) is now also unnecessary, and yarn 4 relies on `corepack` instead of the `yarnPath` in the `.yarnrc.yml`

* [`.yarnrc.yml`](diffhunk://#diff-88fbe28c4102501b94961511a0d70ff895bf39970b4d3fc11917794a239117c5R1-R9): The `plugins` and `yarnPath` have been removed
* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L134-R134): The `packageManager` has been updated from `yarn@3.2.3` to `yarn@4.1.1`.

Documentation update:

* [`docs/tutorials/yarn-migration.md`](diffhunk://#diff-8765f23ed9286e3c311a77236337cd3272635af84764c27bb826ca2db1bffe56L8-R8): The Yarn version in the migration guide has been updated from 3 to 4. The installation instructions have been simplified to only include the command to set the Yarn version. [[1]](diffhunk://#diff-8765f23ed9286e3c311a77236337cd3272635af84764c27bb826ca2db1bffe56L8-R8) [[2]](diffhunk://#diff-8765f23ed9286e3c311a77236337cd3272635af84764c27bb826ca2db1bffe56L44-R44)

References:
https://yarnpkg.com/migration/guide

https://yarnpkg.com/blog/release/4.0#installing-yarn

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by alper98** [comment]
We should inform the integrators that in order to set Backstage up, corepack needs to be enabled
https://yarnpkg.com/blog/release/4.0

**Comment by alper98** [comment]
Some of the workflows fail due to a new setting in the `yarn4` called `Hardened Mode`, which is detected automatically in public PRs. Im not sure if this should be enabled or disabled in the `.yarnrc.yml`, as it comes with some performance drawbacks.

Whats your take @Rugvip  @vinzscam ?

Here's more information: https://yarnpkg.com/blog/release/4.0#hardened-mode

**Comment by taliesins** [comment]
I raised this PR https://github.com/janus-idp/backstage-showcase/pull/1145 to upgrade from Yarn 1 to Yarn 4. 

I didn't run into issues with hardened mode. I think you need to update all the yarn lock files and then you should be good.

Basically I went all in with:
```
yarn upgrade-interactive
# upgraded almost everything, you can delete the yarn file to recalculate everything as well
yarn install
yarn build
```

**Comment by alper98** [comment]
> I raised this PR [janus-idp/backstage-showcase#1145](https://github.com/janus-idp/backstage-showcase/pull/1145) to upgrade from Yarn 1 to Yarn 4.
> 
> I didn't run into issues with hardened mode. I think you need to update all the yarn lock files and then you should be good.
> 
> Basically I went all in with:
> 
> ```
> yarn upgrade-interactive
> # upgraded almost everything, you can delete the yarn file to recalculate everything as well
> yarn install
> yarn build
> ```

It's fantastic to see you've also embraced the upgrade journey! 🚀

I'm a bit curious, though – could you enlighten me on why the yarn upgrade is necessary? Maybe I overlooked it, but I couldn't seem to find any mention in the migration guide from Yarn Classic to Berry ([Yarn Classic to Berry Migration Guide](https://yarnpkg.com/migration/guide)) or even in the release notes for v4 ([Yarn v4 Release Notes](https://yarnpkg.com/blog/release/4.0)).

File: .github/workflows/verify_microsite.yml [hunk 1]
```diff
@@ -35,6 +35,9 @@ jobs:
         with:
           node-version: 18.x
 
+      - name: enable corepack
+        run: corepack enable
```
Comment (Rugvip) [inline .github/workflows/verify_microsite.yml:39]: Super nit but I find the name on these to be a bit redundant, thinking `- run: corepack enable` is enough

**Review by Rugvip — CHANGES_REQUESTED** [review]
Sweet, thank you! 👍 🎉 

Regarding hardened mode, let's try to make it work, I think the additional checks are worth the performance tradeoff

File: docs/tutorials/yarn-migration.md [hunk 2]
```diff
@@ -36,18 +36,40 @@ And this to `.dockerignore`:
...
+
+```bash
+nvm install 18
+```
+
+and then
+
+```bash
+nvm use 18
+```
```
[the remark this answers is not in the document]
Comment (alper98) [reply docs/tutorials/yarn-migration.md:60]: For simplicity, i'll remove it

File: yarn.lock
[the remark this answers is not in the document]
Comment (alper98) [reply yarn.lock:1]: Hmm, not sure. I ran `yarn install` as the only thing. No `yarn upgrade` or anything similar.

Perhaps checking out `yarn.lock` from master, and running yarn install again will do the trick?

**Comment by alper98** [comment]
@Rugvip not sure about the custom regarding resolving threads, but feel free to give a second review.

As of the failing CI, im not sure how to fix this resolution error, do you have any ideas?

File: yarn.lock
[the remark this answers is not in the document]
Comment (Rugvip) [reply yarn.lock:1]: Yep worth a shot, although you'd want to make sure your branch is up to date for that

**Comment by Rugvip** [comment]
Opened #23926 to bump csstype, let's see if that works

**Comment by alper98** [comment]
I updated the css like in https://github.com/backstage/backstage/pull/23926 and ran `yarn install`.

It seems the massive changes to the yarn.lock is, from what i have seen by looking it trough, just a new checksum and a new syntax. The ranges are still the same.

But do double check this

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for this @alper98, great to see things staying up to date. 👍 

I wanted to do this with the Demo site and used this PR to help. Based on that I've left some comments that should help improve things

**Comment by alper98** [comment]
Still active. Have had a busy week in the office. 

Will be back working on this in the coming week

**Comment by alper98** [comment]
> Thanks for this @alper98, great to see things staying up to date. 👍 
> 
> 
> 
> I wanted to do this with the Demo site and used this PR to help. Based on that I've left some comments that should help improve things

Awesome to hear, will have a look 🫡

**Comment by alper98** [comment]
Apologies, very busy week here. I hope i can finish this very soon. If anyone wants to chip in, feel free

**Pull request closed without merging**